app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt

```diff
@@ -1,26 +1,77 @@
 package compiler.intermediate_form
 
+import compiler.ast.Function
+import compiler.ast.Type
 import compiler.ast.Variable
 
 data class FunctionDetailsGenerator(
     val depth: Int,
     val vars: Map<Variable, Boolean>,
-    val parameters: List<Variable>
+    val parameters: List<Variable>,
+    val functionCFG: ControlFlowGraph,
+    val function: Function
 )
+
 data class FunctionCallIntermediateForm(
     val callGraph: ControlFlowGraph,
     val result: IntermediateFormTreeNode?
 )
 
-fun generateCall(args: List<IntermediateFormTreeNode>): FunctionCallIntermediateForm {
-    return TODO()
+val FUNCTION_RESULT_REGISTER = Register()
+
+fun generateCall(fdg: FunctionDetailsGenerator, args: List<IntermediateFormTreeNode>): FunctionCallIntermediateForm {
```

Comment:
Please also move all the functions here to the body of `FunctionDetailsGenerator` and unify the naming convention with [this branch](https://github.com/TCS-Compilers/compiler/blob/9dacb21c97c425af69d3ef695ea3c45cf9977f4f/app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt#L42) before you write tests. This will help with future merging conflicts.

Reply:
Moved and added simple unit test. It's hard to test `genCall` without testing `genRead` and `genWrite`, because we cannot extend `FunctionDetailsGenerator` and override these methods to something dummy for such test class. How about using external mocking library (eg. *Mockito*) in our project to faciliate such testing?

Reply:
That's a great idea. I pushed a commit adding Mockito to the dependencies, you should be able to use it now.